---
title: "Comparing the Direction and Degree of Selection on Mutations Between SARS-CoV-2 Phylogenetic Clades"
output: html_document
date: "`r format(Sys.Date(), '%Y-%m-%d')`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)

#install.packages("dplyr")
#install.packages("ggplot2")
#install.packages("ggforce")
#install.packages("RColorBrewer")
#install.packages("stringr")
#install.packages("gtools")

library(dplyr)
library(ggplot2)
library(ggforce)
library(RColorBrewer)
library(stringr)
library(gtools)

`%notin%` <- Negate(`%in%`)
```

The purpose of this notebook is to generate a volcano plot comparing any two SARS-CoV-2 phylogenetic clades. Using the volcano plot, nucleotide mutations or amino acid mutations at certain sites along the genome that are undergoing very different directions or degrees of selection between the two clades can be identified. Once this notebook has been finalized, sections of the code will be saved as scripts and run through a Snakemake pipeline, allowing all clade comparisons to be computed in parallel.

Listed below are the general steps the notebook uses to generate a volcano plot:\
1. Read in the SARS-CoV-2 data containing observed and expected mutation counts.\
2. Set parameters for how to compute comparisons, including what clades to compare and what countries to accept data from.\
3. Filter data based on parameters.\
4. Restructure data such that there are columns for clade 1 variables and clade 2 variables.\
5. Using contingency table analyses (e.g., Fisher's Exact test), calculate a p-value for each mutation at every nucleotide or amino acid site between the two clades.\
6. Adjust p-values to account for the large number of hypothesis tests.\
7. Using the ratio of observed counts to expected counts for each clade and a pseudocount, calculate log2 fold-change.\
8. Generate a volcano plot with log2 fold-change on the x-axis and -log10 p-values on the y-axis.\
9. Verify the notebook is computing p-values correctly by generating plots with expected outcomes (e.g., synonymous vs non-synonymous mutations).\

First, read in the comma-delimited data from the 'kfeldmann' computational notebook in the Bloom Lab folder on the Fred Hutch server using access to the server or from the Bloom Lab github.com.

```{r read_data}
# Read data using access to the Fred Hutch server.
SARS_mutation <- read.csv("../results/expected_vs_actual_mut_counts/expected_vs_actual_mut_counts.csv", sep=",", header=TRUE)

# Read data from github.com.
#SARS_mutation <- read.csv("https://media.githubusercontent.com/media/jbloomlab/SARS2-mut-rates/main/results/expected_vs_actual_mut_counts/expected_vs_actual_mut_counts.csv", sep=",", header=TRUE)
```

## Define Filtering and Analysis Parameters

```{r parameters}
# The subset column describes what countries the sequence data came from. Identify what the subset options are.
unique(SARS_mutation$subset)

# Define what subset of the data to analyze.
country <- "all"

# Print the clades that are in the data set (i.e., what clades are possible to include in comparisons).
unique(SARS_mutation$clade)

# Identify the two clades that will be included in comparisons.
clades <- c("22A","22B")

# Define a minimum threshold for the number of expected counts. Mutations below this threshold will be removed.
min_expected <- 0

# To avoid infinities when calculating fold-change, add a pseudocount to the observed and expected counts.
pseudo <- 0.5

# Analyze nucleotide mutations or amino acid mutations ('nt' or 'aa')?
option <- "aa"
merge_column <- paste(option, "mutation", sep="_")
```

## Filter the SARS-CoV-2 Mutation Data

For the analyses conducted in this notebook, subset the data by the country (countries) the sequence data came from. Additionally, remove low-quality mutations or branches with abnormally high mutation rates.

```{r filter_all_data}
# Include data from certain countries AND exclude branches with very high mutation rates and questionable mutations.
filtered_SARS_mutation <- subset(SARS_mutation, subset==country & exclude=="False" & expected_count>=min_expected)
```

Create nucleotide mutation and amino acid mutation data sets. Actual and expected counts for rows with identical amino acid mutations are added together. Both data sets only include columns relevant for that mutation to make understanding restructuring the data and analyses easier.

```{r group_mutation}
# Group by amino acid mutation and add counts for identical amino acid mutations.
# NOTE: Grouping with and without 'gene' creates a different number of rows because some equivalent aa mutations belong to different genes. There are 2,160 aa mutations that have duplicates.
aa_group <- as.data.frame(filtered_SARS_mutation %>% group_by(clade, aa_mutation, clade_founder_aa, codon_site, mutant_aa, synonymous, gene) %>% summarise(actual_count=sum(actual_count), expected_count=sum(expected_count)))
  
# Reduce columns for nucleotide mutation analyses to make workflow easier to follow.
nt_group <- filtered_SARS_mutation[,c("clade","nt_mutation","clade_founder_nt","nt_site","synonymous","four_fold_degenerate","gene","actual_count","expected_count")]

# consider entire gene column
# Should I compare mutations with different wild types? (A200T vs C200T) Plot?
```

## Describe and Characterize the Data

Take a look at the first 5 rows of the data.

```{r head_data}
head(get(paste(option, "group", sep="_")), n=5)
```

Look at a summary of the data:

-What are the data types for each column?\
-Is there a similar number of TRUEs and FALSEs for logical variables?\
-Describe the numeric variables (e.g., minimum value).\

```{r summary_data}
summary(get(paste(option, "group", sep="_")))
```

Take a closer look at how many nucleotide sites, amino acid sites and the proportion of mutations in the data set.

```{r describe_data}
# How many nucleotide sites?
length(unique(nt_group$nt_site))

# How many amino acid sites?
length(unique(aa_group$codon_site))

# What percentage of the data set is each type of nucleotide mutation? (e.g., A -> T)
nt_type <- paste(str_sub(nt_group$nt_mutation,1,1), ".", str_sub(nt_group$nt_mutation,-1,-1), sep="")
(table(nt_type)/length(nt_type))*100

# What percentage of the data set is each type of amino acid mutation? (e.g., K -> M)
aa_type <- paste(aa_group$clade_founder_aa, ".", aa_group$mutant_aa, sep="")
(table(aa_type)/length(aa_type))*100
```

## Visualize the Data

After numerically describing the nucleotide and amino acid mutation data, use stacked bar plots to visual the proportion of each mutation for each clade. The number of mutations is normalized by the expected count.

```{r stacked_nucleotide, echo=FALSE}
stacked_nt <- nt_group
stacked_nt$mutation_type <- paste(str_sub(nt_group$nt_mutation,1,1), ".", str_sub(nt_group$nt_mutation,-1,-1), sep="")

ggplot(stacked_nt, aes(x=clade, y=actual_count/expected_count, fill=mutation_type))+
  geom_bar(position="stack", stat="identity")+
  scale_fill_brewer(palette="Paired", name="Nucleotide Mutation")+
  theme_bw()+
  labs(x="Phylogenetic Clade", y="Normalized Mutation Count")
```

In the following amino acid mutation plot, individual plots indicate the starting amino acid (i.e., clade founder) and stacked bars within each plot indicate what that amino acid mutated to.

```{r stacked_amino_acid, echo=FALSE}
stacked_col <- colorRampPalette(brewer.pal(8,"Set1"))(21)

stacked_aa <- aa_group
stacked_aa$clade_founder <- str_sub(aa_group$clade_founder_aa,1,1)
stacked_aa$mutant <- str_sub(aa_group$mutant_aa,1,1)

ggplot(stacked_aa, aes(x=clade, y=actual_count/expected_count, fill=mutant))+
  geom_bar(position="stack", stat="identity")+
  facet_wrap(~clade_founder, ncol=7)+
  scale_fill_manual(values=stacked_col, name="Amino Acid Mutation")+
  scale_x_discrete(guide=guide_axis(n.dodge=5))+
  theme_bw()+
  labs(x="Phylogenetic Clade", y="Normalized Mutation Count")
```

## Prepare and Restructure Data for Analyses

```{r rename_variable}
# Rename variable for analyses to make code more readable.
filtered_SARS_mutation <- get(paste(option, "group", sep="_"))
```

To compare mutations between phylogenetic clades, this notebook will plot volcano plots with -log10 p-value on the y-axis and log2 fold-change on the x-axis. Prior to filtering the data for contingency table analyses, calculate the ratio of observed to expected counts and add a pseudocount to prevent infinities when fold-change is calculated. Mutations with a small count (e.g., 1) are going to be affected by the pseudocount more than mutations with a large count (e.g., 100).

```{r pseudocount}
# Calculate the ratio of actual counts (i.e., observed) to expected counts (i.e., expected) with a pseudocount.
filtered_SARS_mutation$observed_expected <- (filtered_SARS_mutation$actual_count + pseudo)/(filtered_SARS_mutation$expected_count + pseudo)
```

Contingency table analyses can only accept integers, so round the expected counts prior to running analyses (i.e., gives control over how values are rounded).

```{r round_expected}
# Round expected counts because contingency table analyses (e.g., Fisher's Exact test) can only analyze integers.
filtered_SARS_mutation$rounded_expected <- round(filtered_SARS_mutation$expected_count)
```

To quickly calculate the p-value for each contingency table analysis and ensure the user can easily understand what is being compared in each hypothesis test, restructure the data such that all of the variables necessary to run the statistical test are in one row. That includes observed and expected counts for clade 1 and observed and expected counts for clade 2.

```{r restructure_data}
# Create two data frames with data for clade 1 and clade 2.
clade1 <- filtered_SARS_mutation[which(filtered_SARS_mutation$clade==clades[1]),]
clade2 <- filtered_SARS_mutation[which(filtered_SARS_mutation$clade==clades[2]),]

# Append '1' or '2' to the column names so that the two data frames can be easily combined.
colnames(clade1)[colnames(clade1) %notin% merge_column] <- paste(colnames(clade1)[colnames(clade1) %notin% merge_column], "1", sep="_")
colnames(clade2)[colnames(clade2) %notin% merge_column] <- paste(colnames(clade2)[colnames(clade2) %notin% merge_column], "2", sep="_")

# Merge the two data frames by mutation so that each row represents one comparison.
comparisons <- merge(clade1, clade2, by=merge_column)

# Determine if any of the columns from 'clade1' and 'clade2' are the same.
col_names <- colnames(filtered_SARS_mutation)[colnames(filtered_SARS_mutation) %notin% merge_column]
is_equal <- data.frame(sapply(col_names, FUN=function(c){identical(comparisons[,paste(c[1], 1, sep="_")], comparisons[,paste(c[1], 2, sep="_")])}))
col_equal <- row.names(is_equal)[which(is_equal[1]=="TRUE")]

# Rename one of the duplicate columns and remove the other.
colnames(comparisons)[colnames(comparisons) %in% paste(col_equal, "1", sep="_")] <- col_equal
comparisons[,paste(col_equal, "2", sep="_")] <- NULL
```

## Contingency Table Analyses

For each comparison, calculate the probability of obtaining the observed and expected counts assuming there is no difference in selection on mutations between the clades being compared (i.e., p-value) using the Fisher's Exact test. 

```{r contingency_table}
# Iterate through rows and compare clade 1 observed and expected values to clade 2 observed and expected values.
p_values <- data.frame(p_value=apply(comparisons, 1, FUN=function(c) {fisher.test(rbind(c(as.numeric(c["actual_count_1"]),as.numeric(c["rounded_expected_1"])),c(as.numeric(c["actual_count_2"]),as.numeric(c["rounded_expected_2"]))))$p.value}))

# Save and load the variable because calculating p-values may take a long time.
#save(p_values, file=paste("../results/expected_and_actual_counts_ratio/pvalues", paste(clades, collapse="_"), option, ".Rda", sep=""))
load(paste("../results/expected_and_actual_counts_ratio/pvalues", paste(clades, collapse="_"), option, ".Rda", sep=""))
```

Combine the p-values calculated using the contingency table analyses with the data frame of comparisons.

```{r combine_p_values}
# Combine the p-value list with the comparison list.
comparison_p_values <- cbind(comparisons, p_values)
```

## P-Value Correction

By calculating a p-value for every nucleotide or amino acid mutation, comparing the two clades generates a lot of hypothesis tests. Account for this large number of hypothesis tests by correcting the p-values using a false discovery rate correction.

```{r p_value_correction}
# Using a false discovery rate correction, adjust the p-values by the number of hypothesis tests.
comparison_p_values$p_value_corrected <- p.adjust(comparison_p_values$p_value, method="fdr", n=nrow(comparison_p_values))
```

## Volcano Plot

Calculate fold-change using the observed-expected ratios calculated with a pseudocount. Fold-change is calculated as log2 of the clade 1 ratio divided by the clade 2 ratio. Therefore, if fold-change is positive then the clade 1 ratio is greater than the clade 2 ratio, and vise versa if fold-change is negative. 

```{r fold-change}
# Calculate fold-change for plotting volcano plots.
comparison_p_values$fold_change <- log2(comparison_p_values$observed_expected_1/comparison_p_values$observed_expected_2)
```

Generate volcano plots with -log10 p-values on the y-axis and log2 fold-change on the x-axis:

The first volcano plot uses the uncorrected p-values, and the red line indicates a significance value of P = 0.05.

```{r volcano_all, echo=FALSE}
# Plot the uncorrected p-values.
ggplot(comparison_p_values, aes(x=fold_change, y=-log10(p_value)))+
  geom_point(alpha=0.5)+
  geom_hline(yintercept=-log10(0.05), col="red", linetype=2)+
  theme_bw()
```

After correcting the p-values, a large portion of the p-values equal one. The second volcano plot drops P = 1 and uses the remaining corrected p-values. Points are colored by the starting nucleotide or amino acid mutation (i.e., clade founder) and labels indicate the mutation. The red line indicates a significance value of P = 0.05.

```{r volcano_corrected, echo=FALSE}
# After dropping P = 1, plot the corrected p-values.
volcano <- comparison_p_values[which(comparison_p_values$p_value_corrected!=1),]
if("gene" %notin% colnames(volcano)){
  volcano$gene <- ifelse(volcano$gene_1==volcano$gene_2, volcano$gene_1, paste(volcano$gene_1, volcano$gene_2, sep=";"))
}
color <- colorRampPalette(brewer.pal(8,"Set1"))(length(unique(volcano$gene)))

p <- ggplot(volcano, aes(x=fold_change, y=-log10(p_value_corrected), label=get(merge_column), color=gene))+
  geom_point()+
  geom_text(color="black", size=2, nudge_y=0.2)+
  geom_hline(yintercept=-log10(0.05), col="red", linetype=2)+
  scale_color_manual(values=color, name="mutation")+
  theme_bw()

p + theme(legend.position="none")
```

```{r volcano_corrected_legend}
p + theme(legend.position="bottom")
```

## Verify Analyses

To verify the analyses are generating expected results, compare the corrected p-values for synonymous mutations to non-synonymous mutations. Unlike non-synonymous mutations, synonymous mutations should not be experiencing selective pressures. Therefore, while there might be differences in the direction and degree of selection on non-synonymous (i.e., p-values close to 0), there should be insignificant differences between clades for synonymous mutations (i.e., p-values close to 1). The black line indicates a significance value of P = 0.05.

```{r verify_analyses, echo=FALSE}
# Plot the corrected p-values for synonymous and non-synonymous mutations.
verify_analyses <- comparison_p_values
if("synonymous" %notin% colnames(verify_analyses)){
  verify_analyses$synonymous <- ifelse(verify_analyses$synonymous_1=="True" & verify_analyses$synonymous_2=="True", "True", "False")
}
verify_analyses$p_one <- ifelse(verify_analyses$p_value_corrected==1, "P=1", "P<1")

p <- ggplot(verify_analyses, aes(x=p_value_corrected, fill=synonymous, label=after_stat(count)))+
  geom_histogram()+
  geom_text(size=2, vjust=-0.5, stat="bin", position="stack")+
  geom_vline(data=data.frame(xint=0.05, p_one="P<1"), aes(xintercept=xint), linetype=2)+
  scale_x_continuous(breaks=c(0,0.25,0.5,0.75,1))+
  labs(x="corrected p-values")+
  theme_bw()
  
p + ggforce::facet_row(vars(p_one), scales="free", space="free")
```