---
title: "Comparing the Direction and Degree of Selection on Mutations Between SARS-CoV-2 Phylogenetic Clades"
output: html_document
date: "`r format(Sys.Date(), '%Y-%m-%d')`"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)

#install.packages("data.table")
#install.packages("ggplot2")
#install.packages("ggforce")
#install.packages("stringr")
#install.packages("gtools")

library(data.table)
library(ggplot2)
library(ggforce)
library(stringr)
library(gtools)

`%notin%` <- Negate(`%in%`)
```

The purpose of this notebook is to generate a volcano plot comparing any two SARS-CoV-2 phylogenetic clades. Using the volcano plot, mutations at certain nucleotide sites that are undergoing very different directions or degrees of selection between the two clades can be identified. Once this notebook has been finalized, sections of the code will be saved as scripts and run through a Snakemake pipeline, allowing all clade comparisons to be computed in parallel.

Listed below are the general steps the notebook uses to generate a volcano plot:\
1. Read in the SARS-CoV-2 data containing observed and expected mutation counts.\
2. Set parameters for how to compute comparisons, including what clades to compare and what countries to accept data from.\
3. Filter data based on parameters.\
4. Restructure data such that there are columns for clade 1 variables and clade 2 variables.\
5. Using contingency table analyses (e.g., Fisher's Exact test), calculate p-values for each mutation at every nucleotide site between the two clades.\
6. Adjust p-values to account for the large number of hypothesis tests.\
7. Using the ratio of observed counts to expected counts for each clade and a pseudocount, calculate log2 fold-change.\
8. Generate a volcano plot with log2 fold-change on the X-axis and -log10 p-values on the Y-axis.\
9. Verify the notebook is computing p-values correctly by generating plots with expected outcomes (e.g., synonymous mutations).\

First, read in the comma-delimited data from the 'kfeldmann' computational notebook in the Bloom Lab folder on the Fred Hutch server using access to the server or from the Bloom Lab github.com.

```{r read_data}
# Read data using access to the Fred Hutch server.
SARS_mutation <- fread("../results/expected_vs_actual_mut_counts/expected_vs_actual_mut_counts.csv", sep=",", header=TRUE)

# Read data from github.com.
#SARS_mutation <- fread("https://media.githubusercontent.com/media/jbloomlab/SARS2-mut-rates/main/results/expected_vs_actual_mut_counts/expected_vs_actual_mut_counts.csv", sep=",", header=TRUE)
```

## Define Filtering and Analysis Parameters

```{r parameters}
# The subset column describes what countries the sequence data came from. Identify what the subset options are.
unique(SARS_mutation$subset)

# Define what subset of the data to analyze.
country <- "all"

# Print the clades that are in the data set (i.e., what clades are possible to include in comparisons).
unique(SARS_mutation$clade)

# Identify the two clades that will be included in comparisons.
clades <- c("20A","22B")

# Define a minimum threshold for the number of expected counts. Mutations below this threshold will be removed.
min_expected <- 0

# To avoid infinities when calculating fold-change, add a pseudocount to the observed and expected counts.
pseudo <- 0.5

option <- "aa" # nt or aa
merge_columns <- c("nt_mutation","nt_site","clade_founder_nt","gene","codon_position")
merge_columns <- c("aa_mutation","gene")
```

## Filter and Prepare the SARS-CoV-2 Mutation Data

With the data set, both nucleotide and amino acid mutations can be compared between clades. However, there is no column identifying the amino acid mutation (i.e., starting amino acid, codon position along the genome and mutated amino acid). Create a column with this information using the nucleotide site, codon position (i.e., 1, 2, or 3), clade founder amino acid and mutant amino acid columns.

```{r aa_mutation}
# Create a list with the starting nucleotide site for an amino acid for all rows in the data set.
starting_nt <- SARS_mutation$nt_site-as.numeric(str_sub(SARS_mutation$codon_position,1,1))+1

# Create column with amino acid mutation (e.g., M.266.268.L)
SARS_mutation$aa_mutation <- paste(str_sub(SARS_mutation$clade_founder_aa,1,1), starting_nt, starting_nt+2, str_sub(SARS_mutation$mutant_aa,1,1), sep=".")
```

For the analyses conducted in this notebook, subset the data by the country (countries) the sequence data came from. Additionally, remove low-quality mutations or branches with abnormally high mutation rates.

```{r filter_all_data}
# Include data from certain countries AND exclude branches with very high mutation rates and questionable mutations.
filtered_SARS_mutation <- subset(SARS_mutation, subset==country & exclude=="FALSE" & expected_count>=min_expected)
```

To compare mutations between phylogenetic clades, this notebook will plot volcano plots with -log10 p-value on the Y-axis and log2 fold-change on the X-axis. Prior to filtering the data for contingency table analyses, calculate the ratio of observed to expected counts and add a pseudocount to prevent infinities when fold-change is calculated. Mutations with a small count (i.e., 1) are going to be affected by the pseudocount more than mutations with a large count (i.e., 100).

```{r pseudocount}
# Calculate the ratio of actual counts (i.e., observed) to expected counts (i.e., expected) with a pseudocount.
filtered_SARS_mutation$observed_expected <- (filtered_SARS_mutation$actual_count + pseudo)/(filtered_SARS_mutation$expected_count + pseudo)
```

Contingency table analyses can only accept integers, so round the expected counts prior to running analyses (i.e., gives control over how values are rounded).

```{r round_expected}
# Round expected counts because contingency table analyses (e.g., Fisher's Exact test) can only analyze integers.
filtered_SARS_mutation$rounded_expected <- round(filtered_SARS_mutation$expected_count)
```

```{r clean_columns}
# Remove columns that were already used to filter the data.
filtered_SARS_mutation$subset <- NULL
filtered_SARS_mutation$exclude <- NULL
```

## Describe and Characterize the Data

Take a look at the first 5 rows of the data.

```{r head_data}
head(filtered_SARS_mutation, n=5)
```

Look at a summary of the data.\
-What are the data types for each column?\
-Is there a similar number of TRUEs and FALSEs for logical variables?\
-Describe the numeric variables (e.g., minimum value).\

```{r summary_data}
summary(filtered_SARS_mutation)
```

Take a closer look at how many nucleotide sites and the proportion of mutations in the data set.

```{r describe_data}
# How many nucleotide sites?
length(unique(filtered_SARS_mutation$nt_site))

# What percentage of the data set is each type of mutation? (e.g., A -> T)
mutation_type <- paste(str_sub(filtered_SARS_mutation$nt_mutation,1,1), ".", str_sub(filtered_SARS_mutation$nt_mutation,-1,-1), sep="")
(table(mutation_type)/length(mutation_type))*100
```

## Visualize the Data

The four nucleotides (e.g., A) and their possible mutations (e.g., A -> T) may be under different directions and degrees of selection between clades leading to an unequal distribution of nucleotides in the genome. Look at how many mutations (normalized by the expected count) there are for each clade.

```{r stacked_nucleotide, echo=FALSE}
stacked_nt <- filtered_SARS_mutation
stacked_nt$mutation_type <- paste(str_sub(filtered_SARS_mutation$nt_mutation,1,1), ".", str_sub(filtered_SARS_mutation$nt_mutation,-1,-1), sep="")

ggplot(stacked_nt, aes(x=clade, y=actual_count/expected_count, fill=mutation_type))+
  geom_bar(position="stack", stat="identity")+
  scale_fill_brewer(palette="Paired")+
  theme_bw()+
  labs(x="Phylogenetic Clade", y="Normalized Mutation Count")
```

Amino acids occur at different proportions throughout the SARS-CoV-2 genome and between phylogenetic clades. [INSERT]

```{r stacked_amino_acid, echo=FALSE}
stacked_aa <- filtered_SARS_mutation
stacked_aa$clade_founder <- str_sub(filtered_SARS_mutation$clade_founder_aa,1,1)
stacked_aa$mutant <- str_sub(filtered_SARS_mutation$mutant_aa,1,1)

ggplot(stacked_aa, aes(x=clade, y=actual_count/expected_count, fill=mutant))+
  geom_bar(position="stack", stat="identity")+
  facet_wrap(~clade_founder, ncol=7)+
  theme_bw()+
  labs(x="Phylogenetic Clade", y="Normalized Mutation Count")
```

## Restructure Data for Analyses

With such a large data set, it could take hours to calculate the p-values for all clade comparisons. To decrease the amount of time it takes to conduct analyses, this notebook will only compare two clades at a time and the script can be run in parallel to generate all comparisons at once. Therefore, filter the data to only include mutation data from the two clades being compared.

```{r subset_by_clade}
# Create a subset of the data using only the clades that are going to be compared.
filtered_clades <- filtered_SARS_mutation[filtered_SARS_mutation$clade %in% clades,]
```

To quickly calculate the p-value for each contingency table analysis and ensure the user can easily understand what is being compared in each hypothesis test, restructure the data such that all of the variables necessary to run the statistical test are in one row. That includes observed and expected counts for clade 1 and observed and expected counts for clade 2.

```{r restructure_data}
# Create two data frames with data for clade 1 and clade 2.
clade1 <- filtered_clades[which(filtered_clades$clade==clades[1]),]
clade2 <- filtered_clades[which(filtered_clades$clade==clades[2]),]

# Append 'clade1' or 'clade2' to the column names so that the two data frames can be combined with easy to interpret columns.
colnames(clade1)[colnames(clade1) %notin% merge_columns] <- paste(colnames(clade1)[colnames(clade1) %notin% merge_columns], "1", sep="_")
colnames(clade2)[colnames(clade2) %notin% merge_columns] <- paste(colnames(clade2)[colnames(clade2) %notin% merge_columns], "2", sep="_")

# Merge the two data frames by mutation so that each row represents one comparison.
comparisons <- merge(clade1, clade2, by=merge_columns)
```

## Contingency Table Analyses

For each comparison, calculate the probability of obtaining the observed and expected counts assuming there is no difference in selection on mutations between the clades being compared (i.e., p-value) using the Fisher's Exact test. 

```{r contingency_table}
# Iterate through rows and compare clade 1 observed and expected values to clade 2 observed and expected values.
#p_values <- data.frame(p_value=apply(comparisons, 1, FUN=function(c) {fisher.test(rbind(c(as.numeric(c["actual_count_1"]),as.numeric(c["rounded_expected_1"])),c(as.numeric(c["actual_count_2"]),as.numeric(c["rounded_expected_2"]))))$p.value}))

# Save and load variable because calculating p-values may take a long time.
#save(p_values, file=paste("pvalues", paste(clades, collapse="_"), option, ".Rda", sep=""))
load(paste("pvalues", paste(clades, collapse="_"), option, ".Rda", sep=""))
```

Combine the p-values calculated using the contingency table analyses with the data frame of comparisons.

```{r combine_p_values}
# Combine the p-value list with the comparison list.
comparison_p_values <- cbind(comparisons, p_values)
```

## P-Value Correction

```{r p_value_correction}
# Using False Discovery Rate Correction, adjust the p-values by the number of hypothesis tests.
comparison_p_values$p_value_corrected <- p.adjust(comparison_p_values$p_value, method="fdr", n=nrow(comparison_p_values))
```

## Volcano Plot

Calculate fold-change using the observed-expected ratios calculated with a pseudocount. Fold-change is calculated as log2 of the clade 1 ratio divided by the clade 2 ratio. Therefore, if fold-change is positive then the clade 1 ratio is greater than the clade 2 ratio, and vise versa if fold-change is negative. 

```{r fold-change}
# Calculate fold change for plotting volcano plots.
comparison_p_values$fold_change <- log2(comparison_p_values$observed_expected_1/comparison_p_values$observed_expected_2)
```

Generate a volcano plot with -log10 p-values on the Y-axis and log2 fold-change on the X-axis.

```{r volcano_all, echo=FALSE}
# Plot the uncorrected p-values.
ggplot(comparison_p_values, aes(x=fold_change, y=-log10(p_value)))+
  geom_point(alpha=0.5)+
  geom_hline(yintercept=-log10(0.05), col="red", linetype=2)+
  theme_bw()
```

```{r volcano_corrected, echo=FALSE}
# After dropping p = 1, plot FDR corrected p-values (color by nucleotide clade founder and label by mutation).
volcano <- comparison_p_values[which(comparison_p_values$p_value_corrected!=1),]
ggplot(volcano, aes(x=fold_change, y=-log10(p_value_corrected), label=get(paste(option, "mutation", sep="_")), color=str_sub(get(paste(option, "mutation", sep="_")),1,1)))+
  geom_point()+
  geom_text(color="black", size=2, nudge_y=0.2)+
  geom_hline(yintercept=-log10(0.05), col="red", linetype=2)+
  #scale_color_brewer(palette="Set2", name="mutation")+
  theme_bw()
```

## Verify Analyses

To verify the analyses are generating expected results, compare the corrected p-values for synonymous mutations to non-synonymous mutations. Unlike non-synonymous mutations, synonymous mutations should not be experiencing selective pressures. Therefore, while there might be differences in the direction and degree of selection on non-synonymous (i.e., p-values close to 0), there should be insignificant differences between clades for synonymous mutations (i.e., p-values close to 1).

```{r verify_analyses}
# Plot the corrected p-values for synonymous and non-synonymous mutations and compare to a significance value of 0.05.
verify_analyses <- comparison_p_values
verify_analyses$synonymous <- ifelse(verify_analyses$synonymous_1=="TRUE" & verify_analyses$synonymous_2=="TRUE", "TRUE", "FALSE")
verify_analyses$p_one <- ifelse(verify_analyses$p_value_corrected==1, "P=1", "P<1")

p <- ggplot(verify_analyses, aes(x=p_value_corrected, fill=synonymous, label=..count..))+
  geom_histogram()+
  geom_text(size=2, vjust=-0.5, stat="bin", position="stack")+
  geom_vline(data=data.frame(xint=0.05, p_one="P<1"), aes(xintercept=xint), linetype=2)+
  scale_x_continuous(breaks=c(0,0.25,0.5,0.75,1))+
  labs(x="FDR corrected p-values")+
  theme_bw()
  
p + ggforce::facet_row(vars(p_one), scales="free", space="free")
```